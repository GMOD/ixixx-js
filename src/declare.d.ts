declare module "big-map-simple";
