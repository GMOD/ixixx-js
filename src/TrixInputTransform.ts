import { Transform } from 'stream'

export class TrixInputTransform extends Transform {
  _transform(chunk: Buffer, _encoding: unknown, done: () => void) {
    const [id, ...words] = chunk.toString().split(/\s+/)

    this.push(words.map(word => `${word.toLowerCase()} ${id}\n`).join(''))
    done()
  }
}
