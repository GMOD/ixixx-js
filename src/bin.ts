#!/usr/bin/env node
import { ixIxx } from "./index";

const [file, out1 = "out.ix", out2 = "out.ixx"] = process.argv.slice(2);

(async () => {
  await ixIxx(file, out1, out2);
})();
