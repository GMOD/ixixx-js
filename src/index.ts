import { promisify } from "util";
import { finished, Readable, PassThrough, Transform } from "stream";

import pump from "pump";
import split2 from "split2";
import { once } from "events";
import fs from "fs";
import readline from "readline";
import tmp from "tmp";
import esort from "external-sorting";

tmp.setGracefulCleanup();

const streamFinished = promisify(finished); // (A)

// this file (ixixx.ts) is a translation of ixIxx.c from ucscGenomeBrowser/kent
// the license of that file is reproduced below

/*
 * MIT License
 *
 * Copyright (C) 2001 UC Regents
 *
 * Permission is hereby granted, free of charge, to any person or non-commercial
 * entity obtaining a copy of this software and associated documentation files (the
 * "Software"), to deal in the Software without restriction, including without
 * limitation the rights to use, copy, modify, merge, publish, distribute,
 * sublicense, and/or sell copies of the Software, and to permit persons to whom
 * the Software is furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
 * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
 * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
 * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
 * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

const prefixSize = 5;
let binSize = 64 * 1024;

// Characters that may be part of a word
const wordMiddleChars = [] as boolean[];
const wordBeginChars = [] as boolean[];

function isalpha(c: string) {
  return (c >= "a" && c <= "z") || (c >= "A" && c <= "Z");
}

function isdigit(c: string) {
  return c >= "0" && c <= "9";
}

function isalnum(c: string) {
  return isalpha(c) || isdigit(c);
}

function initCharTables() {
  for (let c = 0; c < 256; ++c) {
    if (isalnum(String.fromCharCode(c))) {
      wordBeginChars[c] = wordMiddleChars[c] = true;
    }
  }
  wordBeginChars["_".charCodeAt(0)] = wordMiddleChars["_".charCodeAt(0)] = true;
  wordMiddleChars[".".charCodeAt(0)] = true;
  wordMiddleChars["-".charCodeAt(0)] = true;
}

class TrixInputTransform extends Transform {
  _transform(chunk: Buffer, encoding: any, callback: Function) {
    const [id, ...words] = chunk.toString().split(/\s+/);
    callback(
      null,
      words.map((word) => `${word.toLowerCase()} ${id}\n`).join("")
    );
  }
}

class TrixOutputTransform extends Transform {
  buff = [] as string[];
  current = "";
  _transform(chunk: Buffer, encoding: any, callback: Function) {
    const [id, data] = chunk.toString().split(" ");
    if (this.current !== id) {
      if (this.buff.length) {
        callback(
          null,
          `${this.current} ${this.buff
            .map((elt, idx) => `${elt},${idx + 1}`)
            .join(" ")}\n`
        );
        this.buff = [];
      } else {
        callback(null, null);
      }
      this.current = id;
    } else {
      callback(null, null);
    }
    this.buff.push(data);
    console.log("transform", { len: this.buff });
  }
  _flush(callback: Function) {
    console.log("final", { curr: this.current, len: this.buff });
    if (this.buff.length) {
      callback(
        null,
        `${this.current} ${this.buff
          .map((elt, idx) => `${elt},${idx + 1}`)
          .join(" ")}\n`
      );
    }
    callback(null, null);
  }
}

async function makeIxStream(fileStream: Readable, outIxFilename: string) {
  initCharTables();

  const tmpdir = tmp.dirSync({
    prefix: "jbrowse-trix-sort",
  });

  const out = fs.createWriteStream(outIxFilename);

  // see https://stackoverflow.com/questions/68835344/ for explainer of writer
  const r = split2();
  let ret = pump(r, new TrixOutputTransform(), out);
  console.log("her3");
  await esort({
    //@ts-ignore
    input: pump(fileStream, split2(), new TrixInputTransform()),
    //@ts-ignore
    output: r,
    tempDir: tmpdir.name,
  }).asc();
  console.log("her2");

  await streamFinished(ret);
  console.log("her1");
}

async function makeIx(inFile: string, outIndex: string) {
  const fileStream = fs.createReadStream(inFile);
  return makeIxStream(fileStream, outIndex);
}

function getPrefix(word: string) {
  return word.slice(0, prefixSize).padEnd(5, " ");
}

async function makeIxx(inIx: string, outIxx: string) {
  console.log("her1");
  const out = fs.createWriteStream(outIxx);
  try {
    const fileStream = fs.createReadStream(inIx);
    const rl = readline.createInterface({
      input: fileStream,
    });

    let lastPrefix;
    let writtenPrefix;
    let writtenPos = -binSize;
    let startPrefixPos = 0;
    let bytes = 0;

    for await (const line of rl) {
      const [word] = line.split(/\s/);
      const curPrefix = getPrefix(word);
      if (curPrefix !== lastPrefix) {
        startPrefixPos = bytes;
      }

      if (bytes - writtenPos >= binSize && curPrefix !== writtenPrefix) {
        const res = out.write(
          `${curPrefix}${startPrefixPos
            .toString(16)
            .toUpperCase()
            .padStart(10, "0")}\n`
        );

        // Handle backpressure
        // ref https://nodesource.com/blog/understanding-streams-in-nodejs/
        if (!res) {
          await once(out, "drain");
        }
        writtenPos = bytes;
        writtenPrefix = curPrefix;
      }
      lastPrefix = curPrefix;
      bytes += line.length + 1;
    }
  } finally {
    out.end();
    await streamFinished(out);
  }
}

export async function ixIxx(inText: string, outIx: string, outIxx: string) {
  await makeIx(inText, outIx);
  await makeIxx(outIx, outIxx);
}

export async function ixIxxStream(
  stream: Readable,
  outIx: string,
  outIxx: string
) {
  await makeIxStream(stream, outIx);
  await makeIxx(outIx, outIxx);
}
