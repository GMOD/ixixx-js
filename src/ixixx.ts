import fs from "fs";
import { open } from "fs/promises";
import readline from "readline";
/* Characters that may be part of a word. */

const wordMiddleChars = [] as boolean[];
const wordBeginChars = [] as boolean[];

function isalpha(c: string) {
  return (c >= "a" && c <= "z") || (c >= "A" && c <= "Z");
}

function isdigit(c: string) {
  return c >= "0" && c <= "9";
}

function isalnum(c: string) {
  return isalpha(c) || isdigit(c);
}

function initCharTables() {
  /* Initialize tables that describe characters. */
  for (let c = 0; c < 256; ++c) {
    if (isalnum(String.fromCharCode(c))) {
      wordBeginChars[c] = wordMiddleChars[c] = true;
    }
  }
  wordBeginChars["_".charCodeAt(0)] = wordMiddleChars["_".charCodeAt(0)] = true;
  wordMiddleChars[".".charCodeAt(0)] = true;
  wordMiddleChars["-".charCodeAt(0)] = true;
}

function skipToWord(s: string) {
  /* Skip to next word character.  Return NULL at end of string. */
  let i = 0;
  while (i < s.length) {
    if (wordBeginChars[s.charCodeAt(i)]) {
      return i;
    }
    i += 1;
  }
  return s.length;
}

/* Skip to next non-word character.  Returns empty string at end. */
function skipOutWord(start: string) {
  const s = start;
  let i = 0;
  while (i < s.length) {
    if (!wordMiddleChars[s.charCodeAt(i)]) {
      break;
    }
    i++;
  }

  // while (s > start && !wordBeginChars[(int)(s[-1])])
  //     s -= 1;
  return i;
}

interface WordPos {
  /* Word position. */
  next: WordPos /* Next wordPos in list. */;
  itemId: string /* ID of associated item.  Not allocated here*/;
  wordIx: number /* Word number within doc. */;
}

function wordPosCmp(va: WordPos, vb: WordPos): number {
  /* Compare two wordPos by itemId. */
  const a = va;
  const b = vb;
  let dif = a.itemId.localeCompare(b.itemId);
  if (dif == 0) {
    dif = a.wordIx - b.wordIx;
  }
  return dif;
}

type Hash = any;

function indexWords(
  wordHash: Hash,
  itemId: string,
  words: string[],
  itemIdHash: Hash
) {
  itemIdHash[itemId] = true;
  words.forEach((word, wordIx) => {
    if (!wordHash[word]) {
      wordHash[word] = { name: word, val: [] };
    }
    wordHash[word].val.push({ itemId, wordIx });
  });
}

async function writeIndexHash(wordHash: any, fileName: string) {
  let els = Object.values(wordHash) as any[];
  const file = await open(fileName, "w");
  console.log({ file });

  els.forEach(({ name, val }) => {
    const entries = val.map((pos: any) => `${pos.itemId},${pos.wordIx}`);
    file.writeFile(`${name} ${entries.join(" ")}\n`);
  });
  file.close();
}

async function makeIx(inFile: string, outIndex: string) {
  initCharTables();
  /* Create an index file. */
  const fileStream = fs.createReadStream(inFile);
  const rl = readline.createInterface({
    input: fileStream,
    output: process.stdout,
  });

  const wordHash = {};
  const itemIdHash = {};

  for await (const line of rl) {
    const [id, ...text] = line.toLowerCase().split(/\s/);
    indexWords(
      wordHash,
      id,
      text.filter((f) => !!f),
      itemIdHash
    );
  }

  writeIndexHash(wordHash, outIndex);
}

const prefixSize = 5;
let binSize = 2 ^ 14;
function getPrefix(word: string) {
  return word.slice(0, prefixSize).padEnd(5, " ");
}

type File = any;
function writeIxxEntry(file: File, prefix: string, pos: number) {
  file.writeFile(`${prefix}${pos}\n`);
}

async function makeIxx(inIx: string, outIxx: string) {
  const fileStream = fs.createReadStream(inIx);
  const rl = readline.createInterface({
    input: fileStream,
    output: process.stdout,
  });
  const outFile = open(outIxx, "w");

  let lastPrefix;
  let writtenPos = 0;
  let startPrefixPos = 0;

  let bytes = 0;
  const it = rl[Symbol.asyncIterator]();
  const { value: line } = await it.next();

  const [word] = line.split(/\s/);
  let writtenPrefix = getPrefix(word);
  writeIxxEntry(outFile, writtenPrefix, writtenPos);

  bytes += line.length;
  writtenPos += bytes;

  // const [word, ...rest] = line.split(/\s/);
  // const writtenPrefix = getPrefix(word);
  // const lastPrefix = writtenPrefix;
  // let writtenPos = lineFileTell(lf);
  // writeIxxEntry(f, writtenPrefix, writtenPos);
  /* Loop around adding to index as need be */
  for await (const line of rl) {
    let diff;
    let curPos = bytes;
    // let word = nextWord(line);
    const [word] = line.split(/\s/);
    const curPrefix = getPrefix(word);
    if (curPrefix !== lastPrefix) {
      startPrefixPos = curPos;
    }
    diff = curPos - writtenPos;
    if (diff >= binSize) {
      if (curPrefix !== writtenPrefix) {
        writeIxxEntry(outFile, curPrefix, startPrefixPos);
        writtenPos = curPos;
        writtenPrefix = curPrefix;
      }
    }
    lastPrefix = curPrefix;
    bytes += line.length;
  }
  // carefulClose(&f);
  // lineFileClose(&lf);
  // freeMem(curPrefix);
  // freeMem(lastPrefix);
  // freeMem(writtenPrefix);
}

/* ixIxx - Create indices for simple line-oriented file of format
 * <symbol> <free text>. */
export async function ixIxx(inText: string, outIx: string, outIxx: string) {
  await makeIx(inText, outIx);
  await makeIxx(outIx, outIxx);
}
