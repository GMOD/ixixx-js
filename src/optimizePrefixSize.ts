import fs from 'fs'
import readline from 'readline'
import { binSize, getPrefix } from './util'

export async function optimizePrefixSize(inIx: string) {
  let binSizeTotal = 0
  let binCount = 0
  let prefixSize = 5
  for (; prefixSize < 40; prefixSize++) {
    const fileStream = fs.createReadStream(inIx)
    const rl = readline.createInterface({
      input: fileStream,
    })

    let lastPrefix
    let writtenPrefix
    let writtenPos = -binSize
    let startPrefixPos = 0
    let bytes = 0

    let lastBin = 0
    let maxBinSize = 0

    for await (const line of rl) {
      const [word] = line.split(/\s/)
      const curPrefix = getPrefix(word, prefixSize)
      if (curPrefix !== lastPrefix) {
        startPrefixPos = bytes
      }

      if (bytes - writtenPos >= binSize && curPrefix !== writtenPrefix) {
        const binSize = startPrefixPos - lastBin
        binSizeTotal += binSize
        maxBinSize = Math.max(binSize, maxBinSize)
        binCount++
        lastBin = startPrefixPos

        writtenPos = bytes
        writtenPrefix = curPrefix
      }
      lastPrefix = curPrefix
      bytes += line.length + 1
    }
    const avgBinSize = binSizeTotal / binCount
    // some heuristics. note: binSizeTotal===0 means everything was lumped into
    // one bin
    if (
      (binSizeTotal === 0 && bytes > binSize) ||
      avgBinSize > 3 * binSize ||
      maxBinSize > 10 * binSize
    ) {
      continue
    } else {
      break
    }
  }
  return prefixSize
}
