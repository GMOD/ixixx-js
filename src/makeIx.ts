import { pipeline, Readable } from 'stream'

import split2 from 'split2'
import fs from 'fs'
import { spawn } from 'child_process'
import { TrixInputTransform } from './TrixInputTransform'
import { TrixOutputTransform } from './TrixOutputTransform'

// Characters that may be part of a word
const wordMiddleChars = [] as boolean[]
const wordBeginChars = [] as boolean[]

function isalpha(c: string) {
  return (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z')
}

function isdigit(c: string) {
  return c >= '0' && c <= '9'
}

function isalnum(c: string) {
  return isalpha(c) || isdigit(c)
}

function initCharTables() {
  for (let c = 0; c < 256; ++c) {
    if (isalnum(String.fromCharCode(c))) {
      wordBeginChars[c] = true
      wordMiddleChars[c] = true
    }
  }
  wordBeginChars['_'.charCodeAt(0)] = wordMiddleChars['_'.charCodeAt(0)] = true
  wordMiddleChars['.'.charCodeAt(0)] = true
  wordMiddleChars['-'.charCodeAt(0)] = true
}

export async function makeIxStream(
  fileStream: Readable,
  outIxFilename: string,
) {
  return new Promise((resolve, reject) => {
    initCharTables()

    const out = fs.createWriteStream(outIxFilename)

    // see https://stackoverflow.com/questions/68835344/ for explainer of
    // writer
    const input = pipeline(
      fileStream,
      split2(),
      new TrixInputTransform(),
      err => {
        if (err) {
          reject(err)
        }
      },
    )

    const output = split2()
    pipeline(output, new TrixOutputTransform(), out, err => {
      if (err) {
        reject(err)
      }
    })

    // override locale to C, but keep other env vars
    const sort = spawn('sort', ['-k1,1'], {
      env: { ...process.env, LC_ALL: 'C' },
    })

    input.pipe(sort.stdin)
    sort.stdout.pipe(output)
    output.on('finish', () => {
      resolve(true)
    })
    output.on('error', err => {
      if (err) {
        reject(err)
      }
    })
  })
}

export async function makeIx(inFile: string, outIndex: string) {
  return makeIxStream(fs.createReadStream(inFile), outIndex)
}
