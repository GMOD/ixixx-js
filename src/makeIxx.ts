import { promisify } from 'util'
import { finished } from 'stream'
import { once } from 'events'
import fs from 'fs'
import readline from 'readline'

// locals
import { binSize, getPrefix } from './util'
import { optimizePrefixSize } from './optimizePrefixSize'

const ADDRESS_SIZE = 10

const streamFinished = promisify(finished) // (A)

export async function makeIxx(
  inIx: string,
  outIxx: string,
  prefixSizeParam?: number,
) {
  const out = fs.createWriteStream(outIxx)
  const prefixSize = prefixSizeParam ?? (await optimizePrefixSize(inIx))

  try {
    const fileStream = fs.createReadStream(inIx)
    const rl = readline.createInterface({
      input: fileStream,
    })

    let lastPrefix
    let writtenPrefix
    let writtenPos = -binSize
    let startPrefixPos = 0
    let bytes = 0

    for await (const line of rl) {
      const [word] = line.split(/\s/)
      const curPrefix = getPrefix(word, prefixSize)
      if (curPrefix !== lastPrefix) {
        startPrefixPos = bytes
      }

      if (bytes - writtenPos >= binSize && curPrefix !== writtenPrefix) {
        const res = out.write(
          `${curPrefix}${startPrefixPos
            .toString(16)
            .toUpperCase()
            .padStart(ADDRESS_SIZE, '0')}\n`,
        )

        // Handle backpressure
        // ref https://nodesource.com/blog/understanding-streams-in-nodejs/
        if (!res) {
          await once(out, 'drain')
        }
        writtenPos = bytes
        writtenPrefix = curPrefix
      }
      lastPrefix = curPrefix
      bytes += line.length + 1
    }
  } finally {
    out.end()
    await streamFinished(out)
  }
}
