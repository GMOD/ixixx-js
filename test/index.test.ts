import { ixIxxStream } from '../src/'
import tmp from 'tmp'
import fs from 'fs'
import { Readable } from 'stream'

test('test simple', async () => {
  const l1 = tmp.fileSync()
  const l2 = tmp.fileSync()
  const mytext = Readable.from([
    'heart r222222222\n',
    'brain r333333333\n',
    'kidney r111111111\n',
    'leg r111111111\n',
    'toenail r111111111\n',
    'scalp r9999\n',
  ])
  await ixIxxStream(mytext, l1.name, l2.name)
  const r1 = fs.readFileSync(l1.name, 'utf8')
  const r2 = fs.readFileSync(l2.name, 'utf8')
  expect(r1).toMatchSnapshot()
  expect(r2).toMatchSnapshot()
})

test('volvox sans descriptions', async () => {
  const l1 = tmp.fileSync()
  const l2 = tmp.fileSync()
  const mytext = fs.createReadStream(require.resolve('./volvox.txt'))

  await ixIxxStream(mytext, l1.name, l2.name)
  const r1 = fs.readFileSync(l1.name, 'utf8')
  const r2 = fs.readFileSync(l2.name, 'utf8')
  expect(r1).toMatchSnapshot()
  expect(r2).toMatchSnapshot()
})

test('volvox with descriptions', async () => {
  const l1 = tmp.fileSync()
  const l2 = tmp.fileSync()

  const mytext = fs.createReadStream(
    require.resolve('./volvox_with_descriptions.txt'),
  )

  await ixIxxStream(mytext, l1.name, l2.name)
  const r1 = fs.readFileSync(l1.name, 'utf8')
  const r2 = fs.readFileSync(l2.name, 'utf8')
  expect(r1).toMatchSnapshot()
  expect(r2).toMatchSnapshot()
})
