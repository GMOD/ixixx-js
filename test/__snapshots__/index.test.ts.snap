// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test 1`] = `
"r111111111 kidney,1 leg,2
r222222222 heart,1
r333333333 brain,1
"
`;

exports[`test 2`] = `
"r11110000000000
"
`;
